### ISSUE TITLE
Non-determinsim prevents caching of cc compilation using bazel disk cache

### ISSUE THREAD

**Issue by EricCousineau-TRI** [RobotLocomotion/drake-ros#116]
Relates to #96

It seems like something either in our `bazel_ros2_rules`, or something else, is non-deterministic that prevents cache hits.

First make sure you have `--disk_cache` and `--repository_cache` configured.
The following script can configure it automatically, if you prefer: https://gist.github.com/EricCousineau-TRI/ca97ccc404ece082885d71c559d83936

Then see that multiple clean+rebuild still cause things like `talker_listener.cc` to be recompiled:
```sh
$ cd drake-ros
$ git log -n 1 --oneline --no-decorate
416cf8b Rename not-shimmed executables (#112)
$ git diff  # for reproducibility
diff --git a/ros2_example_bazel_installed/WORKSPACE b/ros2_example_bazel_installed/WORKSPACE
index d390251..d3af953 100644
--- a/ros2_example_bazel_installed/WORKSPACE
+++ b/ros2_example_bazel_installed/WORKSPACE
@@ -40,6 +40,7 @@ ros2_archive(
     # TODO(hidmic,cottsay): Make release-pinned snapshots of this repository once `drake-ros`
     # itself has versions.
     url = "http://repo.ros2.org/ci_archives/drake-ros-underlay/ros2-humble-linux-focal-amd64-ci.tar.bz2",
+    sha256 = "b7a8982cdc07a090b9719e35ae4735df82f9f648d5b703b93bc9db51e448f81d",
     sha256_url = "https://repo.ros2.org/ci_archives/drake-ros-underlay/ros2-humble-linux-focal-amd64-ci-CHECKSUM",
     strip_prefix = "ros2-linux",
 )
$ cd ros2_example_bazel_installed/
$ bazel clean --expunge --async
$ bazel build //ros2_example_apps:talker_listener_cc_test
...
INFO: Elapsed time: 43.834s, Critical Path: 10.97s
INFO: 150 processes: 5 disk cache hit, 142 internal, 3 linux-sandbox.
INFO: Build completed successfully, 150 total actions
$ bazel clean --expunge --async
$ bazel build //ros2_example_apps:talker_listener_cc_test
...
INFO: Elapsed time: 41.517s, Critical Path: 11.12s
INFO: 150 processes: 5 disk cache hit, 142 internal, 3 linux-sandbox.
INFO: Build completed successfully, 150 total actions
```

\cc @sloretz @IanTheEngineer

**Comment by EricCousineau-TRI** [comment]
My assumption is that something is non-deterministic causing the cache miss; there could be something else, but my money is on this.

**Comment by EricCousineau-TRI** [comment]
Resolved by #128
